### DISCUSSION TITLE
round() function

### DISCUSSION THREAD

**Discussion by desram** [micropython/micropython#13080] [Libraries & Drivers]
Hello,
I have a question concerning the "round()" function. round(3.25,1) should return 3.3 but it returns 3.2. Why ?

**Comment by davefes** [comment]
Try 3.26

**Comment by jimmo** [comment]
FYI this matches CPython:

```
Python 3.11.6 (main, Nov 14 2023, 09:36:21) [GCC 13.2.1 20230801] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> round(3.25,1)
3.2
```

e.g. pyboard (single-precision float)

```
MicroPython v1.22.0-preview.118.g841422817e on 2023-11-03; PYBv1.1 with STM32F405RG
Type "help()" for more information.
>>> round(3.25,1)
3.2
>>> 
```

e.g. micropython on linux (double-precision float)

```
MicroPython v1.22.0-preview.164.gfce8d9fd55 on 2023-11-27; linux [GCC 13.2.1] version
Use Ctrl-D to exit, Ctrl-E for paste mode
>>> round(3.25,1)
3.2
```

Are you aware of Python's rounding rules (which is round-to-even rather than round-up for 0.5). Also be aware of floating point precision issues. Please see https://docs.python.org/3/library/functions.html#round for more details.

**Reply by scruss** [reply]
although:

```
Python 3.11.6 (main, Nov  2 2023, 04:51:19) [Clang 14.0.0 (clang-1400.0.29.202)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> round(2.675, 2)
2.67
```
is not the same as
```
MicroPython v1.21.0 on 2023-10-06; Raspberry Pi Pico with RP2040
Type "help()" for more information.
>>> round(2.675, 2)
2.68
```
MicroPython's doing what I'd expect (round-to-even), full-fat Python's off in its own wee world

**Reply by ricksorensen** [reply]
see note in python documentation (pointed to by @jimmo)on round- referring to internal representation of floating point and impact on rounding.
on my windows 10, python 3.8
```
a=2.675
print(f"{a:33.30f}")
2.674999999999999822364316059975
```

**Comment by sosi-deadeye** [comment]
I guess you expected financial rounding. You don't get this in programming languages because the error is bigger, if you sum up financial rounded values.

The scientific rounding is used in programming languages.
Example:

```python
def financial_round(value):
    fraction = value - int(value)
    if fraction >= 0.5:
        return int(value) + 1
    else:
        return int(value)


def gen_seq():
    for x in range(1, 11):
        yield x + 0.5


x = sum(gen_seq())
y = sum(round(x) for x in gen_seq())
z = sum(financial_round(x) for x in gen_seq())

diff1 = abs(x - y)
diff2 = abs(x - z)

```

**Comment by desram** [comment]
Hi,
Well, thanks all of you for this comments.
To davefes: I shall never check a result of a computation in case of giving 3.25 to change it to 3.26 to finally have a right rounding; no way.
To sosi-deadeye: I agree that what I need is what you call "financial-rounding" but to me, "rounding" should be just that.
For information, I re-def the round function with the following code:

def round(val, nbDec):
    f = 10**nbDec
    r = int((val * f)+0.5)/f  # rounding
    if nbDec < 1: # just to return an integer in case of rounding before the decimal dot.
        return int(r)
    else:
        return r

which gives good results in all cases, even with negative nbDec (Number of decimals).

In fact, the problem is not there. 
Indeed, with the native function the execution of :
print(round(3.5, 0) returns 4, which is correct
print(round(3,25, 1) returns 3.2, which is incorrect.
So, to me, the native function is incoherent. Maybe it comes from the implemented "rounding rules" as says jimmo but in this case, rules should be updated...

Regards
